File: upload_to_sheet.py
#!/usr/bin/env python3
"""Append rows from new_rows_2025.csv into target Google Sheet via Playwright."""

from __future__ import annotations

import csv
import json
import time
from pathlib import Path

from playwright.sync_api import sync_playwright

SHEET_ID = "13U7JlDKWQlK64mzg4P9AXLs_QXQwkQvoC82x-9zHcrU"
SOURCE_PATH = Path("/workspace/new_rows_2025.csv")
CHECKPOINT_PATH = Path("/workspace/upload_checkpoint.json")

# One row has already been inserted manually in test phase:
# CSV row 2 -> sheet row 120002.
INITIAL_START_ROW = 120003
SKIP_DATA_ROWS = 1
CHUNK_SIZE = 10000


def clean_cell(value: str) -> str:
    return value.replace("\t", " ").replace("\r", " ").replace("\n", " ").strip()


def iter_chunks():
    with SOURCE_PATH.open("r", encoding="utf-8", newline="") as f:
        reader = csv.reader(f)
        next(reader)  # header
        for _ in range(SKIP_DATA_ROWS):
            next(reader)

        chunk: list[list[str]] = []
        for row in reader:
            # Force text treatment in sheet to avoid date auto-conversion (e.g. 12.3.6).
            row[1] = "'" + clean_cell(row[1])
            normalized = [clean_cell(cell) for cell in row]
            chunk.append(normalized)
            if len(chunk) >= CHUNK_SIZE:
                yield chunk
                chunk = []
        if chunk:
            yield chunk


def load_checkpoint() -> dict:
    if CHECKPOINT_PATH.exists():
        return json.loads(CHECKPOINT_PATH.read_text(encoding="utf-8"))
    return {"next_row": INITIAL_START_ROW, "chunks_done": 0, "rows_done": 0}


def save_checkpoint(next_row: int, chunks_done: int, rows_done: int) -> None:
    payload = {
        "next_row": next_row,
        "chunks_done": chunks_done,
        "rows_done": rows_done,
        "updated_at_epoch": int(time.time()),
    }
    CHECKPOINT_PATH.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")


def main() -> None:
    state = load_checkpoint()
    next_row = int(state["next_row"])
    chunks_done = int(state["chunks_done"])
    rows_done = int(state["rows_done"])

    print(f"Resuming upload from row {next_row}, chunks_done={chunks_done}, rows_done={rows_done}")

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(permissions=["clipboard-read", "clipboard-write"])
        page = context.new_page()

        for idx, chunk in enumerate(iter_chunks(), start=1):
            if idx <= chunks_done:
                continue

            tsv_text = "\n".join("\t".join(row) for row in chunk)
            url = f"https://docs.google.com/spreadsheets/d/{SHEET_ID}/edit#gid=0&range=A{next_row}"
            page.goto(url, wait_until="domcontentloaded", timeout=120000)
            page.wait_for_timeout(10000)
            page.keyboard.press("Escape")
            page.wait_for_timeout(300)

            page.evaluate("(txt) => navigator.clipboard.writeText(txt)", tsv_text)
            page.keyboard.press("Control+V")

            # Large multi-row paste requires time for grid update/autosave.
            pause_ms = min(45000, 7000 + len(chunk) * 2)
            page.wait_for_timeout(pause_ms)

            next_row += len(chunk)
            chunks_done += 1
            rows_done += len(chunk)
            save_checkpoint(next_row=next_row, chunks_done=chunks_done, rows_done=rows_done)

            print(
                f"Uploaded chunk {chunks_done}: rows={len(chunk)}, "
                f"rows_done={rows_done}, next_row={next_row}"
            )

        browser.close()

    print("Upload completed.")


if __name__ == "__main__":
    main()

File: upload_to_sheet.js
const fs = require("fs");
const path = require("path");
const { chromium } = require("playwright");

const SHEET_URL =
  "https://docs.google.com/spreadsheets/d/13U7JlDKWQlK64mzg4P9AXLs_QXQwkQvoC82x-9zHcrU/edit?usp=sharing";
const CSV_PATH = path.resolve(process.argv[2] || "lots_2025_final.csv");
const START_CELL = (process.argv[3] || "A1").toUpperCase();
const CLEAR_FIRST = (process.argv[4] || "").toLowerCase() === "clear";

function csvToTsv(csvText) {
  const rows = [];
  let cur = "";
  let row = [];
  let inQuotes = false;

  for (let i = 0; i < csvText.length; i += 1) {
    const ch = csvText[i];
    const next = i + 1 < csvText.length ? csvText[i + 1] : "";

    if (ch === '"') {
      if (inQuotes && next === '"') {
        cur += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
      continue;
    }

    if (!inQuotes && ch === ",") {
      row.push(cur);
      cur = "";
      continue;
    }

    if (!inQuotes && (ch === "\n" || ch === "\r")) {
      if (ch === "\r" && next === "\n") {
        i += 1;
      }
      row.push(cur);
      rows.push(row);
      row = [];
      cur = "";
      continue;
    }

    cur += ch;
  }

  if (cur.length > 0 || row.length > 0) {
    row.push(cur);
    rows.push(row);
  }

  return rows.map((r) => r.join("\t")).join("\n");
}

function splitRows(tsv) {
  const lines = tsv.split("\n");
  const header = lines[0] || "";
  const data = lines.slice(1).filter((x) => x.length > 0);
  return { header, data };
}

async function waitEditor(page) {
  await page.waitForTimeout(5000);
  const candidates = [
    "role=grid",
    "div#waffle-grid-container",
    "div.docs-sheet-container",
  ];
  for (const sel of candidates) {
    try {
      await page.waitForSelector(sel, { timeout: 15000 });
      return;
    } catch (_err) {
      // try next selector
    }
  }
  throw new Error("Google Sheets editor not detected.");
}

async function pasteChunk(page, chunkTsv) {
  await page.keyboard.insertText(chunkTsv);
}

async function goToCellOrRange(page, value) {
  await page.keyboard.press("Control+g");
  await page.waitForTimeout(500);
  await page.keyboard.type(value);
  await page.keyboard.press("Enter");
  await page.waitForTimeout(700);
}

async function run() {
  if (!fs.existsSync(CSV_PATH)) {
    throw new Error(`CSV not found: ${CSV_PATH}`);
  }
  const csv = fs.readFileSync(CSV_PATH, "utf8");
  const tsv = csvToTsv(csv.replace(/^\uFEFF/, ""));
  const { header, data } = splitRows(tsv);

  const browser = await chromium.launch({
    headless: true,
    args: ["--disable-dev-shm-usage"],
  });
  const context = await browser.newContext();
  const page = await context.newPage();

  await page.goto(SHEET_URL, { waitUntil: "domcontentloaded", timeout: 120000 });
  await waitEditor(page);

  if (START_CELL === "A1" && CLEAR_FIRST) {
    // Clear a wide range to remove leftovers from previous uploads.
    await goToCellOrRange(page, "A1:I1000000");
    await page.keyboard.press("Backspace");
    await page.waitForTimeout(1200);
    // Second pass helps remove residual rows left by partial browser operations.
    await page.keyboard.press("Backspace");
    await page.waitForTimeout(1200);
  }

  await goToCellOrRange(page, START_CELL);

  // For chunk-sized files (~100k rows) a single paste is more reliable
  // than iterative sub-pastes with cursor movement.
  const totalRows = data.length + 1;
  const allLines = [header, ...data];
  const payload = allLines.join("\n");
  await pasteChunk(page, payload);

  const approxRows = totalRows;
  const waitMs = Math.min(10 * 60 * 1000, Math.max(30_000, Math.floor(approxRows / 200) * 1000));
  await page.waitForTimeout(waitMs);

  // Trigger save sync
  await page.keyboard.press("Control+s");
  await page.waitForTimeout(8000);

  await context.close();
  await browser.close();
  console.log(`Uploaded ${approxRows} rows from ${CSV_PATH}`);
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});
